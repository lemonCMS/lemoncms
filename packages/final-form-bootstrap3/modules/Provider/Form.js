import React from "react";
import PropTypes from "prop-types";
import { Form as FinalForm } from "react-final-form";
import arrayMutators from "final-form-arrays";
import ContextWrapper from "../Context/ContextWrapper";

const Form = props => {
  const { children, className, debug, readOnly, listen, ...formProps } = props;

  return (
    <FinalForm
      {...formProps}
      mutators={{
        ...arrayMutators
      }}
      render={({ handleSubmit, ...rest }) => {
        return (
          <ContextWrapper
            {...rest}
            debug={debug}
            listen={listen}
            readOnly={readOnly}
          >
            <form onSubmit={handleSubmit} className={className}>
              {children}
            </form>
          </ContextWrapper>
        );
      }}
    />
  );
};

Form.propTypes = {
  className: PropTypes.string,
  validate: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  debug: PropTypes.bool,
  readOnly: PropTypes.bool,
  listen: PropTypes.func,
  initialValues: PropTypes.objectOf(
    PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.array,
      PropTypes.object,
      PropTypes.bool
    ])
  )
};

Form.defaultProps = {
  className: "",
  validate: () => {},
  listen: () => {},
  debug: false,
  readOnly: false,
  initialValues: {}
};

export default Form;
