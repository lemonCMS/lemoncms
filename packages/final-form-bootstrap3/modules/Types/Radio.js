import React from "react";
import PropTypes from "prop-types";
import RadioAlias from "react-bootstrap/lib/Radio";
import context from "../decorators/context";
import fieldGroup from "./fieldGroup";

const createRadioButtons = (children, initialValue, emit, name) => {
  if (
    children &&
    Array.isArray(children) &&
    React.isValidElement(children[0]) &&
    children[0].type === "option"
  ) {
    return children.map(option => {
      const { value, children } = option.props;
      return (
        <RadioAlias
          name={name}
          key={value}
          checked={value === initialValue}
          onChange={event => {
            emit(event, value);
          }}
        >
          {children}
        </RadioAlias>
      );
    });
  }
};

const Radio = props => {
  const { input, children } = props;

  const emit = (event, value) => {
    input.onChange(value);
  };

  return <>{createRadioButtons(children, input.value, emit, input.name)}</>;
};

Radio.propTypes = {
  input: PropTypes.shape({
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onBlur: PropTypes.func.isRequired,
    onFocus: PropTypes.func.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
      PropTypes.number
    ])
  }),
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
  placeholder: PropTypes.string,
  label: PropTypes.string,
  help: PropTypes.string,
  computedInvalid: PropTypes.bool.isRequired
};

Radio.defaultProps = {
  input: {},
  children: [],
  label: null,
  help: null,
  placeholder: null
};

export default context()(fieldGroup(Radio));
