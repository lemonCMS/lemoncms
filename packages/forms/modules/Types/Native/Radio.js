import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import context from '../../decorators/context';
import field from '../../decorators/field';

const Radio = props => {
  const {
    input, children, formControl, computedInvalid
  } = props;

  const radioList = Array.isArray(children) ? children : [children];

  const onChange = (event, key, option) => {
    input.onChange(option.props.value);
  };

  const list = radioList.map((option, key) => (
    <div className="form-check" key={option.props.value}>
      <input
        type="radio"
        name={`${input.name}`}
        id={`${input.name}_${key}`}
        value={option.props.value}
        className={classNames('form-check-label')}
        onChange={event => onChange(event, key, option)}
        checked={String(input.value) === String(option.props.value)}
      />{' '}
      <label htmlFor={`${input.name}_${key}`} className={classNames('form-check-label')}>
        {option.props.children}
      </label>
    </div>
  ));

  return <div className={classNames('ffc-input', { 'is-invalid': computedInvalid }, formControl)}>{list}</div>;
};

Radio.propTypes = {
  input: PropTypes.oneOfType([PropTypes.object]),
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  formControl: PropTypes.string,
  computedInvalid: PropTypes.bool.isRequired
};

Radio.defaultProps = {
  input: {},
  formControl: null
};

export default context(field(Radio));
