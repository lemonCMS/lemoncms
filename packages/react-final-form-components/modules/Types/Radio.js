import React, { Component } from "react";
import ReactDOMServer from "react-dom/server";
import PropTypes from "prop-types";
import memoize from "memoize-one";
import RadioAlias from "react-bootstrap/lib/Radio";
import Col from "react-bootstrap/lib/Col";
import Row from "react-bootstrap/lib/Row";
import context from "../decorators/context";
import fieldGroup from "./fieldGroup";
import Filter from "./components/Filter";

class Radio extends Component {
  state = {
    values: [],
    filterText: ""
  };

  componentDidMount() {
    const {
      input: { value }
    } = this.props;

    if (typeof value !== "undefined") {
      this.setState({ values: value });
    }
  }

  emit = (event, radioValue) => {
    const {
      input: { onChange }
    } = this.props;
    onChange(radioValue);
  };

  filtered = memoize((list, filterText) => {
    return list.filter(item => {
      if (!filterText || item.type !== "option") {
        return true;
      }

      if (typeof item.props.children === "string") {
        return (
          item.props &&
          item.type === "option" &&
          item.props.children.includes(filterText)
        );
      }

      return ReactDOMServer.renderToString(item.props.children).includes(
        filterText
      );
    });
  });

  handleChange = event => {
    this.setState({ filterText: event.target.value });
  };

  clearFilterText = event => {
    this.setState({ filterText: "" });
  };

  createRadios = children => {
    const { columns, filter, input, isDisabled } = this.props;
    if (
      children &&
      Array.isArray(children) &&
      React.isValidElement(children[0]) &&
      children[0].type === "option"
    ) {
      const list = children.map(option => {
        if (option.type !== "option") {
          return option;
        }

        const { value, children } = option.props;
        return (
          <RadioAlias
            key={value}
            disabled={isDisabled}
            checked={input.value === value}
            onChange={event => {
              this.emit(event, value);
            }}
          >
            {children}
          </RadioAlias>
        );
      });

      if (columns === 1) {
        return list;
      }

      const itemsPerColumn = Math.ceil(list.length / columns);
      // create new array of length with undefined values
      const cols = Array.apply(null, Array(columns)).map(() => {});
      const display = cols.map((col, index) => {
        const start = index * itemsPerColumn;
        const end = start + itemsPerColumn;
        return (
          <Col sm={Math.round(12 / columns)} key={index}>
            {list.slice(start, end)}
          </Col>
        );
      });

      return <Row>{display}</Row>;
    }

    // Filtered returned nothing
    if (filter) {
      return <div>No results</div>;
    }

    return (
      <RadioAlias
        checked={input.value}
        disabled={isDisabled}
        onChange={event => {
          this.emit(event, true, true);
        }}
      >
        {children}
      </RadioAlias>
    );
  };

  render() {
    const { filter, placeholderFilter, children, isDisabled } = this.props;
    const { filterText } = this.state;
    if (filter) {
      const filteredList = this.filtered(children, this.state.filterText);
      return (
        <>
          <Filter
            filterText={filterText}
            clearFilterText={this.clearFilterText}
            handleChange={this.handleChange}
            placeholderFilter={placeholderFilter}
            isDisabled={isDisabled}
          />
          {this.createRadios(filteredList)}
        </>
      );
    }

    return <>{this.createRadios(children)}</>;
  }
}

Radio.propTypes = {
  input: PropTypes.shape({
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onBlur: PropTypes.func.isRequired,
    onFocus: PropTypes.func.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
      PropTypes.number,
      PropTypes.array
    ])
  }),
  children: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.string,
    PropTypes.object,
    PropTypes.element
  ]),
  disabled: PropTypes.func,
  isDisabled: PropTypes.bool,
  placeholderFilter: PropTypes.string,
  label: PropTypes.string,
  help: PropTypes.string,
  mandatory: PropTypes.bool,
  columns: PropTypes.number,
  filter: PropTypes.bool,
  computedInvalid: PropTypes.bool.isRequired,
  layout: PropTypes.shape({
    label: PropTypes.object,
    field: PropTypes.object
  })
};

Radio.defaultProps = {
  input: {},
  children: [],
  label: null,
  help: null,
  columns: 1,
  filter: false,
  placeholderFilter: "Filter",
  disabled: null,
  isDisabled: false,
  mandatory: false,
  layout: null
};

export default context()(fieldGroup(Radio));
