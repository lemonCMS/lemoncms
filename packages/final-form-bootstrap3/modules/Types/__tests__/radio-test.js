import renderer from "react-test-renderer";
import { mount } from "enzyme";
import Form from "../../Provider/Form";
import React from "react";
import Radio from "../Radio";

const field = () => (
  <Radio
    name={"field"}
    placeholder={"a nice placeholder"}
    help={"What was your mothers maiden name?"}
  >
    <option value={"1"}>choice 1</option>
    <option value={"2"}>choice 2</option>
    <option value={"3"}>choice 3</option>
    <option value={"4"}>choice 4</option>
  </Radio>
);

describe("BS3 radiobutton", () => {
  it("should render", () => {
    const tree = renderer.create(
      <Form onSubmit={() => {}} initialValues={{ field: "1" }}>
        {field()}
      </Form>
    );
    const output = tree.toJSON();
    expect(output).toMatchSnapshot();
  });

  it("should return single value", () => {
    const callback = jest.fn();

    const component = mount(<Form onSubmit={callback}>{field()}</Form>);

    const inputs = component.find("input");
    expect(inputs).toHaveLength(4);

    inputs
      .at(0)
      .simulate("change", {
        target: { name: "field", value: "1", checked: true }
      });
    inputs
      .at(1)
      .simulate("change", {
        target: { name: "field", value: "2", checked: true }
      });
    inputs
      .at(2)
      .simulate("change", {
        target: { name: "field", value: "3", checked: true }
      });

    const helpBlock = component.find(".help-block");
    expect(helpBlock.text()).toEqual("What was your mothers maiden name?");

    component.find("form").simulate("submit");
    expect(callback).toHaveBeenCalledWith(
      {
        field: "3"
      },
      expect.any(Object),
      expect.any(Function)
    );
  });
});
