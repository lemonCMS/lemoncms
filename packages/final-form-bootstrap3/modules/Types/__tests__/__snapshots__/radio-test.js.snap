// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BS3 radiobutton should render 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="form-group has-feedback"
  >
    <div
      className="radio"
    >
      <label
        title=""
      >
        <input
          checked={true}
          disabled={false}
          name="field"
          onChange={[Function]}
          type="radio"
        />
        choice 1
      </label>
    </div>
    <div
      className="radio"
    >
      <label
        title=""
      >
        <input
          checked={false}
          disabled={false}
          name="field"
          onChange={[Function]}
          type="radio"
        />
        choice 2
      </label>
    </div>
    <div
      className="radio"
    >
      <label
        title=""
      >
        <input
          checked={false}
          disabled={false}
          name="field"
          onChange={[Function]}
          type="radio"
        />
        choice 3
      </label>
    </div>
    <div
      className="radio"
    >
      <label
        title=""
      >
        <input
          checked={false}
          disabled={false}
          name="field"
          onChange={[Function]}
          type="radio"
        />
        choice 4
      </label>
    </div>
    <span
      className="help-block"
    >
      What was your mothers maiden name?
    </span>
  </div>
</form>
`;
